### PR TITLE
K disjoint paths null pointer bug fix v2

### PR DESCRIPTION
**Pull request by bkrogh**
Identified a bug causing null pointer exceptions in k disjoint shortest path algorithms. 
Added a minimal test case that triggers the bug in both Suurballe and Bhandari algorithms. 
Implemented bug fixes.
[#790](https://github.com/jgrapht/jgrapht/issues/790)

Fixed formatting issues

----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

File: jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/KDisjointShortestPathsTestCase.java [hunk 1]
```diff
@@ -925,6 +927,76 @@ public void testStar()
         }
     }
 
+    /**
+     * A complex test case with the goal of finding the three shortest paths from vertex 1 to vertex 2 through the following weighted directed graph. Vertices are numbers in boxes, i.e., 1, 3, 4, 5, 6, 7, 8, 2.
+     * Weights are numbers close to an edge. Each edge has its origin to the left, and destination to the right
+     * The source is node 1. Sink is node 2.
+     * The weight of each edge is the unboxed number close the edge.
+     *
```
Comment (jsichi) [inline jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/KDisjointShortestPathsTestCase.java]: Add //@formatter:on and off around this diagram, otherwise it will get destroyed next time we reformat our code.

**Comment by jsichi** [comment]
Thanks for the very clear test case.  I think I understand the logic of the fix (for an even number of overlaps, each path has to go a different way, whereas for an odd number, the odd man out can use the edge), but I'll leave this open for a few days in case @jkinable or @assimiz has comments.

File: jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/KDisjointShortestPathsTestCase.java (hunk 1 above)
Comment (bkrogh) [reply jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/KDisjointShortestPathsTestCase.java]: Thank you for the suggestion.

I added formatter:off/on around the diagram.

**Comment by bkrogh** [comment]
I made three minor changes:
1. Typo fix in comment on test case
2. Added formatter:off/on tags as requested
3. Removed an edge from the graph in the added test case.[1]

[1] This edge was incorrectly included in the original commit. I had just added it temporarily to check a corner case in the algorithm.

**Comment by bkrogh** [comment]
> Thanks for the very clear test case. 
I am glad it is appreciated. It is distilled from a real world occuring graph containing 600+ edges, so it took some time to remove all unnessary edges/vertices, and simplify weights, while still triggering the issue. :)

> I think I understand the logic of the fix (for an even number of overlaps, each path has to go a different way, whereas for an odd number, the odd man out can use the edge)
Exactly. 

I actually first tried to make a simple adaptation of the original code so that it would work correctly with the given test case. It was not possible to do so (I could not figure it out at least).
While it is trivial to count the number of times each edge is used, the issue, in the original implementation, is that it first find duplicate edges, and then excludes all these from the found shortest paths. Because an edge can only be marked duplicate or not duplicate, and there are three shortest paths that use a specific edge in the new test case, this edge is either excluded from the final result or included multiple times. 
Excluding the edge is incorrect, because the 3 disjoint shortest paths require the edge. Including it multiple times is incorrect too, because the disjoint edge shortest paths are not allowed to reuse any edges.

I therefore ended up reimplementing the two methods.

> I'll leave this open for a few days in case @jkinable or @assimiz has comments.
Good idea. Especially considering that this PR reimplements two central shared methods of the two algorithms. As I said, I tried to avoid reimplementing the methods completely, but it turned out not to be possible.
 I am very happy to hear that this is on track for merge, since it prevents us from using the official jgrapht releases.

**Comment by jsichi** [comment]
Thanks!  I'm merging it now, so it will appear in the next SNAPSHOT build, and then it will be in the official release we're working on.
